pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1750,3 +1769,82 @@ func (r *ReconcileClusterDeployment) deleteStaleProvisions(provs []*hivev1.Clust
 		}
 	}
 }
+
+// getAllSyncSetInstances return all syncset instances for a cluster deployment
+func (r *ReconcileClusterDeployment) getAllSyncSetInstances(cd *hivev1.ClusterDeployment) ([]*hivev1.SyncSetInstance, error) {
+
+	// get all syncset instances for the clusterdeployment namespace
+	ssiList := &hivev1.SyncSetInstanceList{}
+	err := r.Client.List(context.TODO(), ssiList, client.InNamespace(cd.Namespace))
+	if err != nil {
+		return nil, err
+	}
+
+	syncSetInstances := []*hivev1.SyncSetInstance{}
+	for i, syncSetInstance := range ssiList.Items {
+		if syncSetInstance.Spec.ClusterDeployment.Name == cd.Name {
+			syncSetInstances = append(syncSetInstances, &ssiList.Items[i])
+		}
+	}
+	return syncSetInstances, nil
+}
+
+// checkFailureSyncCondition return true when syncsetinstance status condition has hivev1.ApplyFailureSyncCondition
+func checkFailureSyncCondition(syncSetInstances []*hivev1.SyncSetInstance) bool {
+	for _, syncSetInstance := range syncSetInstances {
+		for _, condition := range syncSetInstance.Status.Conditions {
+			if condition.Type == hivev1.ApplyFailureSyncCondition {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// addSyncSetFailedCondition returns true when it sets the hivev1.SyncSetFailedCondition
+// hivev1.SyncSetFailedCondition is set if any syncsetinstance has status condition type as hivev1.ApplyFailureSyncCondition will return true
+func (r *ReconcileClusterDeployment) addSyncSetFailedCondition(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (bool, error) {
+
+	// get all syncset instances for this cluster deployment
+	syncSetInstances, err := r.getAllSyncSetInstances(cd)
+	if err != nil {
+		cdLog.WithError(err).Error("unable to list related sync set instances for cluster deployment")
+		return false, err
+	}
+
+	isfailedCondition := checkFailureSyncCondition(syncSetInstances)
+
+	// check if hivev1.SyncSetFailedCondition already set on the cluster deployment
+	currentCond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions, hivev1.SyncSetFailedCondition)
+
+	if currentCond == nil && isfailedCondition == false {
+		return false, nil
+	}
+
+	// add/update hivev1.SyncSetFailedCondition to true
+	if isfailedCondition == true {
+		if currentCond == nil || currentCond.Status == corev1.ConditionFalse {
```

Comment:
With current code `lastTransitionTime:` wont change unless there is a change in condition status. I guess that is fine.

Reply:
That is how `lastTransitionTime` should behave. It is the last time that the condition transitioned from one status to another.